jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DosHandler.java

```diff
@@ -0,0 +1,559 @@
+                {
+                    double elapsedTime1 = (double)(now - _sampleStart);
+                    double count = _sampleCount;
+                    if (elapsedTime1 > 0.0)
+                    {
+                        double currentRate = (count * TimeUnit.SECONDS.toNanos(1L)) / elapsedTime1;
+                        // Adjust alpha based on the ratio of elapsed time to the interval to allow for long and short intervals
+                        double adjustedAlpha = _alpha * (elapsedTime1 / _samplePeriod);
+                        if (adjustedAlpha > 1.0)
+                            adjustedAlpha = 1.0; // Ensure adjustedAlpha does not exceed 1.0
+
+                        _exponentialMovingAverage = (adjustedAlpha * currentRate + (1.0 - adjustedAlpha) * _exponentialMovingAverage);
+                    }
+                    else
+                    {
+                        // assume count as the rate for the sample.
+                        double guessedRate = count * TimeUnit.SECONDS.toNanos(1) / _samplePeriod;
+                        _exponentialMovingAverage = (_alpha * guessedRate + (1.0 - _alpha) * _exponentialMovingAverage);
+                    }
+
+                    // restart the sample
+                    _sampleStart = now;
+                    _sampleCount = 0;
+                }
+
+                // if the rate has been exceeded?
+                return _exponentialMovingAverage > _maxRequestsPerSecond;
+            }
+
+            @Override
+            public boolean isIdle(long now)
+            {
+                return !isRateExceeded(now, false) && _exponentialMovingAverage <= 0.0001;
+            }
+
+            double getCurrentRatePerSecond()
+            {
+                return _exponentialMovingAverage;
+            }
+        }
+    }
+
+    /**
+     * A Handler to reject DoS requests with a status code or failure.
+     */
+    public static class StatusRejectHandler implements Request.Handler
+    {
+        private final int _status;
+
+        public StatusRejectHandler()
+        {
+            this(-1);
+        }
+
+        /**
+         * @param status The status used to reject a request, or 0 to fail the request or -1 for a default ({@link HttpStatus#TOO_MANY_REQUESTS_429}.
+         */
+        public StatusRejectHandler(int status)
+        {
+            _status = status >= 0 ? status : HttpStatus.TOO_MANY_REQUESTS_429;
+        }
+
+        @Override
+        public boolean handle(Request request, Response response, Callback callback) throws Exception
+        {
+            if (_status == 0)
+                callback.failed(new RejectedExecutionException());
+            else
+                Response.writeError(request, response, callback, _status);
+            return true;
+        }
+    }
+
+    /**
+     * A Handler to reject DoS requests after first delaying them.
+     */
+    public static class DelayedRejectHandler extends Handler.Abstract
```

Comment:
Feels like this class could have a `CyclicTimeouts<Exchange>` field that does all the expiration logic -- it should simplify the implementation by a lot and avoid bugs.

Reply:
I had a go at doing this, but I don't think it is right or simpler.  CyclicTimeouts is optimized for timeouts that are expected to be cancelled.  It also does fine grained timeouts that strive to be accurate.    For this delayed handler, we just tick with half the delay period with a timeout that we know will expire and then we process the queue.     I think it is best this way.

Reply:
I fail to understand why it is best having to rewrite the loop, checking for expirations, rescheduling, when it would have been enough to implement `onExpired() { Response.writeError(); }`, leaving all of the above to existing JDK or Jetty utility classes?
If you don't want to use `CyclicTimeouts` just use a plain `Scheduler`, but don't rewrite a scheduler.